### PR TITLE
Windows: Allow enforced ANSI colors for redirected stderr

### PR DESCRIPTION
**Pull request by kinke**
`-color=on` used to be a no-op on Windows if `stdout` wasn't a terminal. Enable ANSI color codes instead, just like on Posix, enabling colored output for GitLab CI for example.

### CODE REVIEW

**Comment by MoonlightSentinel** [comment]
Maybe enable `compilable/testcolour.sh` for windows?

File: src/dmd/console.d
Comment (WalterBright) [inline src/dmd/console.d:27]: add: `else static assert(0); `

**Comment by WalterBright** [comment]
There seems to be a *lot* more going on in this PR than just making -color=on not a no-op.

**Comment by kinke** [comment]
> There seems to be a _lot_ more going on in this PR than just making -color=on not a no-op.

Functionally, only that `detectTerminal()` now checks `stderr` on Windows too (previously `stdout`, but output sent to `stderr`).

As a Windows host may use a `WindowsConsole` (Windows terminal) or `ANSIConsole` (Windows non-terminal + Posix) now, I've moved from heap-allocated host-OS-specific struct to a `Console` interface with 2 class implementations.

I've gotten rid of some Windows-specific code duplication (terminal detection / console creation) I've noticed in the process - I know you don't like mixing refactorings and functional changes, but this is how I work. ;)

**Comment by WalterBright** [comment]
Thanks for the explanation.

I understand this is your style, but it makes it difficult for myself and others to understand your PRs, and causes problems for people trying to track down the source of regressions. We have a lot of experience with PRs that combine minor changes with large refactorings, and it isn't good, which is why the style guidelines are the way they are. If we make exceptions for some people, how can we have this as part of our PR guidelines?

**Comment by WalterBright** [comment]
For example, I have a hard time even identifying in the diffs where the no-op removal is.

**Comment by WalterBright** [comment]
I remember one particularly egregious case where a regression was tracked back to a 3 or 4 hundred line PR, by someone who was no longer around to fix it. With a heavy sigh, I had to sit down and work through the damned thing.

It turned out the bug fix was literally *one line*, and the rest was all refactoring that served to bury and obfuscate it. (The bug fix was erroneous, hence the regression.)

**Comment by kinke** [comment]
I'm not willing to ~double efforts (extra commits with backported logic after noticing ugliness in the midst of a functional revamp) just to please reviewers, especially as we have paid PR managers now.

File: src/dmd/console.d
Comment (kinke) [inline src/dmd/console.d:244]: This ANSI fallback on Windows is the 'no-op change' - Windows-specific `Console.create()` would previously return null if no terminal was detected, disabling fancy colors.

**Comment by WalterBright** [comment]
> I'm not willing to ~double efforts (extra commits with backported logic after noticing ugliness in the midst of a functional revamp) just to please reviewers

As I explained, it is not at all about pleasing reviewers, although it does make reviews easier. It's to save many hours of work for *maintainers*. @atilaneves and I have been having a lot of trouble figuring out the purpose of some of the logic in template instantiations. Going back to the PR for them is rarely helpful, as the fixes were mixed in with voluminous refactorings. A *lot* of time has been wasted dealing with this.

I understand noticing bugs needing fixing while refactoring. Finish the refactor, then fix the bugs. It isn't double the effort, I do it all the time. Or fix the bug, then refactor. Whichever way is easier for the situtation.

Based on https://github.com/dlang/dmd/pull/12541#pullrequestreview-664868625 I infer that the bug fix is *one* line out of nearly 200 lines of change. I recommend reverting that one line, we'll pull this one, and put the one line in a followup PR.

**Comment by WalterBright** [comment]
@wilzbach what do you disagree with?

**Comment by wilzbach** [comment]
I think context matters. This is the CLI coloring, not somewhere deep in semantics.

**Comment by CyberShadow** [comment]
@kinke Have you seen https://github.com/dlang/dmd/pull/8975 ?

**Comment by WalterBright** [comment]
> I think context matters. This is the CLI coloring, not somewhere deep in semantics.

I agree context matters. Now, if in the future it becomes necessary to return to the original behavior, the maintainer is faced with trying to untangle the bug fix from the refactoring. Does this happen? Yes, I've been that maintainer many times. It's much more work to reconstruct what happened and where the fix was, rather than when they are separate PRs. Hence the change to refactorings and bug fixes go in separate PRs, especially when the bug fix is apparently a trivial part of a large refactoring.

One subject per PR.

As for context, it wouldn't be an issue if it was 2 lines of bug fix and five lines of refactoring. Ok. But this is one line of bug fix and 180 lines of refactoring, a clear case of when the bug fix should be separate.

**Review by WalterBright — DISMISSED** [review]
Please put bug fix in separate PR from the refactoring.

**Comment by CyberShadow** [comment]
@WalterBright @kinke I refactored this PR into several smaller commits. It should be much easier to review now.

File: src/dmd/console.d
Comment (CyberShadow) [inline src/dmd/console.d:121]: In my first commit I accidentally replaced `create` with a constructor. The last commit changes it back. A constructor is more idiomatic, I think.

**Comment by kinke** [comment]
Thanks @CyberShadow! No, I wasn't aware of previous attempts (apparently almost identical!); glad you don't need to be convinced anymore and extremely glad you took the time to split it up. :+1:

File: src/dmd/console.d
Comment (kinke) [reply src/dmd/console.d:121]: I just went with/kept the static factory function to have the is-compatible logic and according `return null` inside the class itself; no strong opinion about it though.

Comment (CyberShadow) [reply src/dmd/console.d:121]: The `null` thing is a bit odd. I think it would make more sense to have a no-op `Console` implementation which does nothing. That would allow removing all the `if (con)` checks throughout the codebase elsewhere. But that doesn't have to be done here.

**Comment by RazvanN7** [comment]
@WalterBright are you ok to merge this as is or do you insist on making separate PRs?

**Review by RazvanN7 — APPROVED** [review]

**Comment by WalterBright** [comment]
@RazvanN7 @thewilsonator @kinke  Three commits? I clicked on each one, and each is +177 / -187 changes.

**Comment by kinke** [comment]
4 commits, and the overall line diff stats on the top right on GitHub has been for the whole PR for a while now.

**Comment by WalterBright** [comment]
@kinke I clicked on each commit, and they're all the full commit. There is nothing resembling the one line actual bug fix. We are now in exactly the situation that I wished to avoid - a one line bug fix lost somewhere in a 200 line refactoring.

**Comment by CyberShadow** [comment]
There is no "one line actual bug fix" because the old API was not great and unsuitable to allow one implementation be used on more than one platform.

I think a "one line actual bug fix" would be possible if the changes were arranged into the following sequences of commits:

- Refactor (further split into moving code around, adjusting indentation, etc. as here)
- Introduce new yet-unused redandant code
- "one line actual bug fix" commit
- Delete old no-longer-used redundant code
- Refactor

but that would have been even more work.

**Comment by WalterBright** [comment]
> but that would have been even more work.

I do this all the time. It has never been even more work. Sometimes do the refactor first, then the fix, other times do the fix first, then the refactor. Whichever works better. For this case, the one line fix after the refactoring sounds like it would have worked fine.

Let's say we find this bug fix is wrong for some reason. How do we revert it? This is not a stupid question. I have been faced with this several times - attempting to find out why some several hundred line PR job was failing, only to (after hours of work) discover that the bug fix was exactly *one* line. I am not kidding.

The contributing guidelines clearly says: "if your fix requires a refactoring, then the refactoring should be submitted as a separate pull request."

https://github.com/dlang/dmd/blob/master/CONTRIBUTING.md

@CyberShadow your work in isolating which PR caused a regression has been invaluable. You implicitly know that the first step in fixing a regression is to find which PR introduced it. The larger a PR is, and the more it is full of distracting refactorings, the harder it will be.

Please, everyone, follow the CONTRIBUTING.md guidelines.

**Comment by CyberShadow** [comment]
> Let's say we find this bug fix is wrong for some reason. How do we revert it?

In this particular case, the code is sufficiently simple that simply reading and fixing it is easier than trying black-box mechanical operations such as reverting. Specifically, to revert to the old behavior:

```diff
@@ -236,10 +236,7 @@ bool detectTerminal() nothrow
 Console createConsole(FILE* fp) nothrow
 {
     version (Windows)
-    {
-        if (auto c = WindowsConsole.create(fp))
-            return c;
-    }
+        return WindowsConsole.create(fp);
 
     return new ANSIConsole(fp);
 }
```

However, we should not single out the "one line actual bug fix" as the only change that we would want to revert. Refactorings contain bugs frequently enough too, and, allowing to condense the intended change into a "one line actual bug fix" commit requires ever more refactoring (and therefore more opportunity for bugs).

> Please, everyone, follow the CONTRIBUTING.md guidelines.

I think everyone is on the same page with you. The disconnect is a combination of:

- This code is very non-critical
- This code is sufficiently simple that no experience of working on a compiler is required to adjust it

But also:

- The amount of work required to split a change into smaller ones is much greater than doing it that way from the beginning
- The amount of work required to split a change into smaller ones is much greater than the amount of work required to *review* it

I admit that my refactoring was not perfect (as I mentioned in a review comment above), but I also had to balance the time that I could spend to help here. I'm guessing the author made a similar informed decision in deciding whether to split up the diff while working on it, and reached a different conclusion than you due to a different evaluation of priorities.

BTW I don't mind helping out in such situations such as splitting up large changes, and wouldn't have minded working more on splitting up this PR too upon request, but nobody requested or objected on the current diffs before this was merged.

**Comment by WalterBright** [comment]
> Specifically, to revert to the old behavior:

And **that** code should be in a **followup** PR. This is exactly what I'm talking about. It is **not extra work**.

> we should not single out the "one line actual bug fix" as the only change that we would want to revert.

YES WE SHOULD. Fixing a bug is FUNDAMENTALLY DIFFERENT from introducing a bug.

I am completely astonished by this thread.

> This code is very non-critical

Doesn't matter. Finding and fixing non-critical bugs takes JUST AS MUCH TIME as fixing critical ones. Why, why, WHY make it harder?

> This code is sufficiently simple that no experience of working on a compiler is required to adjust it

Finding those lines in TWO HUNDRED LINES of other random changes is NOT SUFFICIENTLY SIMPLE.

I'M THE ONE WHO GETS THESE BILLS, and I'm sick of it.

This PR should be reverted. Resubmit it with the diff you provided. Then add a followup PR with the diff changes.

**Comment by WalterBright** [comment]
> nobody requested or objected on the current diffs

**I** did. I put in a REQUESTED CHANGES. @thewilsonator dismissed it.

**Comment by WalterBright** [comment]
@kinke your contributions are valuable, and so is your time. I understand that you do not always have the time to adjust your work to fit the contributing guidelines. When this happens, please pass your work along to @RazvanN7 to adjust & submit.

**Pull request merged**
